ctapipe/reco/stereo_combination.py

```diff
@@ -0,0 +1,291 @@
+            if mono.is_valid:
+                values.append(mono.energy.to_value(u.TeV))
+                if tel_id not in event.dl1.tel:
+                    raise ValueError("No parameters for weighting available")
+                weights.append(
+                    self._calculate_weights(event.dl1.tel[tel_id].parameters)
+                )
+                ids.append(tel_id)
+
+        if len(values) > 0:
+            weights = np.array(weights, dtype=np.float64)
+            weights /= weights.max()
+
+            if self.log_target:
+                values = np.log(values)
+
+            mean = np.average(values, weights=weights)
+            std = np.sqrt(np.cov(values, aweights=weights))
+
+            if self.log_target:
+                mean = np.exp(mean)
+                std = np.exp(std)
+
+            valid = True
+        else:
+            mean = std = np.nan
+            valid = False
+
+        event.dl2.stereo.energy[self.prefix] = ReconstructedEnergyContainer(
+            energy=u.Quantity(mean, u.TeV, copy=False),
+            energy_uncert=u.Quantity(std, u.TeV, copy=False),
+            telescopes=ids,
+            is_valid=valid,
+        )
+        event.dl2.stereo.energy[self.prefix].prefix = self.prefix
+
+    def _combine_classification(self, event):
+        ids = []
+        values = []
+        weights = []
+
+        for tel_id, dl2 in event.dl2.tel.items():
+            mono = dl2.classification[self.prefix]
+            if mono.is_valid:
+                values.append(mono.prediction)
+                dl1 = event.dl1.tel[tel_id].parameters
+                weights.append(self._calculate_weights(dl1) if dl1 else 1)
+                ids.append(tel_id)
+
+        if len(values) > 0:
+            mean = np.average(values, weights=weights)
+            valid = True
+        else:
+            mean = np.nan
+            valid = False
+
+        container = ParticleClassificationContainer(
+            prediction=mean, telescopes=ids, is_valid=valid
+        )
+        container.prefix = self.prefix
+        event.dl2.stereo.classification[self.prefix] = container
+
+    def __call__(self, event: ArrayEventContainer) -> None:
+        """
+        Calculate the mean prediction for a single array event.
+        """
+        if self.property == "energy":
+            self._combine_energy(event)
+        elif self.property == "classification":
+            self._combine_classification(event)
+        else:
+            raise NotImplementedError(f"Cannot combine {self.property}")
+
+    def predict_table(self, mono_predictions: Table) -> Table:
+        """
+        Calculates the (array-)event-wise mean.
+        Telescope events, that are nan, get discarded.
+        This means you might end up with less events if
+        all telescope predictions of a shower are invalid.
+        """
+
+        prefix = f"{self.prefix}_tel"
+        # TODO: Integrate table quality query once its done
```

Comment:
Do we want to work on that todo?

Reply:
Yes, but not in this PR, it's too large already.